handler/oauth2/strategy_jwt.go

```diff
@@ -38,6 +38,19 @@ type DefaultJWTStrategy struct {
 	jwt.JWTStrategy
 	HMACSHAStrategy *HMACSHAStrategy
 	Issuer          string
+	ScopeField      jwt.JWTScopeFieldEnum
+}
+
+func (h DefaultJWTStrategy) WithIssuer(issuer string) DefaultJWTStrategy {
+	strategy := h
+	strategy.Issuer = issuer
+	return strategy
+}
```

Comment:
I do not think `With*` should be changing the original value in-place but returning a copy every time.

Reply:
So I am OK with doing:

```
func (h *DefaultJWTStrategy) WithIssuer(issuer string) *DefaultJWTStrategy {
	strategy := *h
	strategy.Issuer = issuer
	return &strategy
}
```

If you prefer that.

Reply:
The problem is that there is no "original" and "copy" - yes it's true for the simple struct values, but anything inside the struct that isn't a simple value but a pointer (so slices, maps, structs, ...) will stay a pointer. So instead of a real copy, you get a pseudo-copy that is half a copy and half the original, which is why it makes sense to just operate on pointer.

Reply:
How I see it is that `WithIssuer` copies only the part it modifies. So that the input does not get modified. Nobody claims that `WithIssuer` returns a copy. What the contract would be is that `WithIssuer` does not modify the input and copies only as little as necessary for that to not happen. Because it is setting a top-level field it just copies the top-level value. If it was setting something deeper, it would have to copy that deeper value, too.

Anyway, it is your library, so I can adapt it they way you want. I just think it is more error prone. I prefer "immutable" coding style here.

Reply:
Ok, looks like `With*` semantics do return a somewhat (url is excluded...) shallow copy in the [Go stdlib](https://golang.org/src/net/http/request.go?s=12980:13039#L341). I think shallow copies are generally a bad idea because of the aforementioned limitations (the stdlib linked suffers from this also). Let me give you an example why I think this interface spec is not great:

```
f := NewSomething()
f.With("bla")
f.Bla // "Something"
```

```
f := NewSomething().With("bla")
f.Bla // "bla"
```

So an inexperienced dev might get this wrong quite easily. But not enough, here's another example:

```
s1 := NewDefaultJWTStrategy()
s2 := s1.WithIssuer("bla")

s2.HMACSHAStrategy.AccessTokenLifespan = time.Minute // modifies s1 also
s2.HMACSHAStrategy = NewHMACStrategy("bla") // does not modify s1
```

As you can see, shallow copies are quite tricky. Since we're configuring a security component here, there should be very little room for error. Maybe the `With*` args are just not the right approach here.

Reply:
OK. As you want. I will make it change things in-place.

Reply:
Just for the record. You yourself use shallow copying in `claims_jwt.go` in `With` and `WithDefaults`.